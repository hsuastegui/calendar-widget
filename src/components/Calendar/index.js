import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import next from "../../assets/next.svg";
import previous from "../../assets/previous.svg";
import {
  getRows,
  getNextMonth,
  getPrevMonth,
  getCurrentMonth,
  getFormattedDate,
  getToday,
  getUpdatedDate,
  format,
} from "../../utils";
import {
  Container,
  Day,
  Bubble,
  Row,
  WeekDays,
  WeekDay,
  Header,
  Button,
} from "./style";

const week = ["Mo", "Tu", "We", "Th", "Fr", "Sa", "Su"];

function Calendar({ open, onSelect, selected }) {
  const [currentMonth, setCurrentMonth] = useState(getCurrentMonth());
  const [selectedDate, setSelectedDate] = useState("");
  const today = getToday();
  const currentDate = getFormattedDate(currentMonth);
  const year = currentDate.year();
  const month = currentDate.format("MMMM");

  useEffect(() => {
    if (selected === "") {
      setSelectedDate("");
      setCurrentMonth(getCurrentMonth());
    }
  }, [selected]);

  const handlePrev = () => {
    setCurrentMonth(getPrevMonth(currentDate));
  };

  const handleNext = () => {
    setCurrentMonth(getNextMonth(currentDate));
  };

  const handleClick = (d) => {
    const date = getUpdatedDate(currentDate, d);
    setSelectedDate(date.format(format));
    onSelect(date.format("DD/MM/YYYY"));
  };

  return (
    <Container open={open}>
      <Header>
        <Button prev onClick={handlePrev}>
          <img src={previous} alt="previous" />
        </Button>
        <span>
          {month} {year}
        </span>
        <Button next onClick={handleNext}>
          <img src={next} alt="next" />
        </Button>
      </Header>
      <WeekDays>
        {week.map((w) => (
          <WeekDay key={w}>{w}</WeekDay>
        ))}
      </WeekDays>
      {getRows(currentDate).map((r) => (
        <Row key={r.join("")}>
          {r.map((d, i) => (
            <Day key={i}>
              {d !== "" && (
                <Bubble
                  onClick={() => {
                    handleClick(d);
                  }}
                  current={today === `${year}-${month}-${d}`}
                  selected={selectedDate === `${year}-${month}-${d}`}
                >
                  {d}
                </Bubble>
              )}
            </Day>
          ))}
        </Row>
      ))}
    </Container>
  );
}

Calendar.defaultProps = {
  open: false,
  onSelect: () => {},
  selected: "",
};

Calendar.propTypes = {
  open: PropTypes.bool,
  onSelect: PropTypes.func,
  selected: PropTypes.string,
};
export default Calendar;
