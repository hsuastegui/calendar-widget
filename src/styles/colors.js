export const brightBlue = "#006cfa";
export const darkishBlue = "#00357c";
export const powderBlue = "#bfdafd";
export const veryLightBlue = "#d9e9fe";
export const white = "#fff";
