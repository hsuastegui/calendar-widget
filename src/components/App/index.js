import React, { useState } from "react";
import upDown from "../../assets/up-down.svg";
import {
  Container,
  Title,
  SecondaryButton,
  Subtitle,
  Button,
  Input,
  InpuText,
  Selector,
} from "../../styles";
import Calendar from "../Calendar";

function App() {
  const [modalOpen, setModalOpen] = useState(false);
  const [selectedDate, setSelectedDate] = useState("");

  const handleModal = () => {
    setModalOpen(!modalOpen);
  };

  const handleSubmit = () => {
    setModalOpen(false);
    setSelectedDate("");
  };

  return (
    <Container>
      <div>
        <Title>Have you sold subject to contract?</Title>
        <Subtitle>
          Enter your exchange date to unlock the tools you need for the next
          stage of your sale.
        </Subtitle>
      </div>
      <Selector>
        <Input>
          <InpuText selected={selectedDate}>
            {selectedDate ? selectedDate : "Please select"}
          </InpuText>
          <SecondaryButton onClick={handleModal}>
            <img src={upDown} alt="up-down" />
          </SecondaryButton>
          <Calendar
            onSelect={setSelectedDate}
            open={modalOpen}
            selected={selectedDate}
          />
        </Input>
        <Button onClick={handleSubmit}>Submit</Button>
      </Selector>
    </Container>
  );
}

export default App;
