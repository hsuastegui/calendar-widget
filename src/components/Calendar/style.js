import styled from "styled-components";
import {
  veryLightBlue,
  brightBlue,
  darkishBlue,
  white,
} from "../../styles/colors";
import { Roboto } from "../../styles/fonts";

export const Container = styled.div`
  width: 392px;
  border-radius: 4px;
  box-shadow: 0 0 32px 0 rgba(0, 53, 124, 0.25);
  border: solid 2px ${darkishBlue};
  background-color: ${white};
  position: absolute;
  top: 46px;
  left: -2px;
  box-sizing: border-box;
  transition: all ease 0.3s;
  overflow: hidden;
  max-height: ${({ open }) => (open ? "500px" : 0)};
  opacity: ${({ open }) => (open ? 1 : 0)};
`;

export const Header = styled.div`
  font-family: ${Roboto};
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  color: ${darkishBlue};
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid rgba(0, 53, 124, 0.25);
`;

export const Button = styled.button`
  background: none;
  border: none;
  cursor: pointer;
  height: 58px;
  width: 42px;
  border-right: ${({ prev }) =>
    prev ? " solid 1px rgba(0, 53, 124, 0.25)" : "none"};
  border-left: ${({ next }) =>
    next ? " solid 1px rgba(0, 53, 124, 0.25)" : "none"};
`;

export const WeekDays = styled.div`
  display: flex;
  border-bottom: solid 1px rgba(0, 53, 124, 0.25);
  padding: 16px 0;
`;

export const WeekDay = styled.span`
  font-family: ${Roboto};
  font-size: 12px;
  font-weight: 700;
  line-height: 1.33;
  color: ${darkishBlue};
  flex: 1 1 14.28%;
  text-align: center;
`;

export const Row = styled.div`
  display: flex;
  border-bottom: solid 1px rgba(0, 53, 124, 0.25);
  padding: 8px 0;
`;

export const Day = styled.div`
  flex: 0 1 14.28%;
`;

export const Bubble = styled.span`
  display: block;
  margin: 0 auto;
  border-radius: 50%;
  cursor: pointer;
  font-family: ${Roboto};
  font-size: 12px;
  line-height: 32px;
  width: 32px;
  text-align: center;
  font-weight: ${({ current, selected }) => (current || selected ? 700 : 400)};
  color: ${({ selected }) => (selected ? white : darkishBlue)};
  background: ${({ selected }) => (selected ? brightBlue : white)};

  &:hover {
    background: ${({ selected }) => (selected ? brightBlue : veryLightBlue)};
    font-weight: 700;
  }
`;
