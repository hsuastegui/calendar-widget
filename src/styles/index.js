import styled from "styled-components";
import { brightBlue, darkishBlue, powderBlue, white } from "./colors";

export const Container = styled.div`
  border-radius: 4px;
  border: solid 2px ${darkishBlue};
  background-color: ${white};
  padding: 46px 136px 46px 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const Title = styled.p`
  font-family: "Roboto", sans-serif;
  font-size: 20px;
  font-weight: 700;
  line-height: 1.4;
  color: ${darkishBlue};
  margin-bottom: 4px;
  margin-top: 0;
`;

export const Subtitle = styled.span`
  font-family: "Roboto", sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  color: ${darkishBlue};
`;

export const Button = styled.button`
  border-radius: 4px;
  background-color: ${brightBlue};
  text-shadow: 1px 1px 1px rgba(0, 28, 64, 0.25);
  font-family: "Roboto", sans-serif;
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  color: ${white};
  width: 99px;
  height: 48px;
  cursor: pointer;
`;

export const SecondaryButton = styled.button`
  background-color: ${powderBlue};
  border: none;
  height: 44px;
  width: 40px;
  border-top-right-radius: 2px;
  border-bottom-right-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  cursor: pointer;
  border-left: 2px solid ${darkishBlue};
`;

export const Input = styled.div`
  width: 245px;
  border-radius: 4px;
  border: solid 2px ${darkishBlue};
  background-color: ${white};
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-left: 15px;
  margin-right: 18px;
  position: relative;
`;

export const InpuText = styled.span`
  opacity: 0.5;
  font-family: "Roboto", sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 44px;
  color: ${darkishBlue};
`;

export const Selector = styled.div`
  display: flex;
  align-items: center;
`;
