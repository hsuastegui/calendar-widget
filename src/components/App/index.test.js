import React from "react";
import { shallow } from "enzyme";
import { SecondaryButton } from "../../styles";
import Calendar from "../Calendar";
import App from "./index";

describe("App", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });
  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("opens calendar", () => {
    wrapper.find(SecondaryButton).simulate("click");
    expect(wrapper.find(Calendar).prop("open")).toBe(true);
  });
});
