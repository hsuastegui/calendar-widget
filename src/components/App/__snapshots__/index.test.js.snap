// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<styled.div>
  <div>
    <styled.p>
      Have you sold subject to contract?
    </styled.p>
    <styled.span>
      Enter your exchange date to unlock the tools you need for the next stage of your sale.
    </styled.span>
  </div>
  <styled.div>
    <styled.div>
      <styled.span
        selected=""
      >
        Please select
      </styled.span>
      <styled.button
        onClick={[Function]}
      >
        <img
          alt="up-down"
          src="up-down.svg"
        />
      </styled.button>
      <Calendar
        onSelect={[Function]}
        open={false}
        selected=""
      />
    </styled.div>
    <styled.button
      onClick={[Function]}
    >
      Submit
    </styled.button>
  </styled.div>
</styled.div>
`;
